src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -118,6 +120,65 @@ internal void DeepInsertRequest<T>(T resource)
             }
         }
 
+        /// <summary>
+        /// Asynchronous deep insert request.
+        /// </summary>
+        /// <typeparam name="T">The type of the top-level object to be deep inserted.</typeparam>
+        /// <param name="resource">The top-level object of the type to be deep inserted.</param>
+        internal void BeginDeepInsertRequest<T>(T resource)
+        {
+            PerRequest pereq = null;
+
+            if (resource == null)
+            {
+                throw Error.ArgumentNull(nameof(resource));
+            }
+
+            BuildDescriptorGraph(this.ChangedEntries, true, resource);
+
+            try
+            {
+                ODataRequestMessageWrapper deepInsertRequestMessage = this.GenerateDeepInsertRequest();
+                this.Abortable = deepInsertRequestMessage;
+
+                if (deepInsertRequestMessage != null)
+                {
+                    deepInsertRequestMessage.SetContentLengthHeader();
+                    this.perRequest = pereq = new PerRequest();
+                    pereq.Request = deepInsertRequestMessage;
+                    pereq.RequestContentStream = deepInsertRequestMessage.CachedRequestStream;
+
+                    AsyncStateBag asyncStateBag = new AsyncStateBag(pereq);
+
+                    this.responseStream = new MemoryStream();
+
+                    IAsyncResult asyncResult = BaseAsyncResult.InvokeAsync(deepInsertRequestMessage.BeginGetRequestStream, this.AsyncEndGetRequestStream, asyncStateBag);
+
+                    pereq.SetRequestCompletedSynchronously(asyncResult.CompletedSynchronously);
+                }
+                else
+                {
+                    this.SetCompleted();
+
+                    if (this.CompletedSynchronously)
+                    {
+                        this.HandleCompleted(pereq);

```

Comment:
Is `HandleCompleted` idempotent? This method would be called here and in the finally block as well...